File: fetch_data.js
#!/usr/bin/env node

/**
 * ScotLIS Property Data Fetcher for Shetland
 *
 * Fetches all property sales from ScotLIS for every valid ZE postcode.
 * Captures ALL titles per address (multiple sales over time).
 *
 * Usage:
 *   node fetch_data.js              # Start or resume fetching
 *   node fetch_data.js --export     # Export cached data to data.js (skip fetching)
 *   node fetch_data.js --stats      # Show progress stats
 *
 * Rate limiting: 18–23 second random delay between requests.
 * Progress is cached in cache/ so it can be resumed if interrupted.
 * Estimated runtime: ~5 hours for all Shetland postcodes.
 */

const https = require("https");
const fs = require("fs");
const path = require("path");

const CACHE_DIR = path.join(__dirname, "cache");
const PROGRESS_FILE = path.join(CACHE_DIR, "progress.json");
const POSTCODES_FILE = path.join(CACHE_DIR, "postcodes.json");
const OUTPUT_FILE = path.join(__dirname, "data.js");

// --- Helpers ---

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function randomDelay() {
  // 18–23 seconds to avoid rate limiting
  return 18000 + Math.random() * 5000;
}

function httpsGet(url) {
  return new Promise((resolve, reject) => {
    const req = https.get(url, { headers: { "User-Agent": "Mozilla/5.0" } }, (res) => {
      let data = "";
      res.on("data", (chunk) => (data += chunk));
      res.on("end", () => {
        if (res.statusCode >= 200 && res.statusCode < 300) {
          try {
            resolve(JSON.parse(data));
          } catch (e) {
            reject(new Error(`JSON parse error for ${url}: ${e.message}`));
          }
        } else {
          reject(new Error(`HTTP ${res.statusCode} for ${url}: ${data.slice(0, 200)}`));
        }
      });
    });
    req.on("error", reject);
    req.setTimeout(30000, () => {
      req.destroy();
      reject(new Error(`Timeout for ${url}`));
    });
  });
}

// --- Postcode Discovery ---

async function fetchAllPostcodes() {
  if (fs.existsSync(POSTCODES_FILE)) {
    console.log("Loading cached postcodes...");
    return JSON.parse(fs.readFileSync(POSTCODES_FILE, "utf8"));
  }

  console.log("Fetching all valid ZE postcodes from postcodes.io...");
  const postcodes = {};
  const prefixes = ["ZE1", "ZE2", "ZE3"];

  for (const prefix of prefixes) {
    // Generate all possible postcode suffixes (0-9 + 0AA-9ZZ)
    for (let num = 0; num <= 9; num++) {
      for (let c1 = 65; c1 <= 90; c1++) {
        for (let c2 = 65; c2 <= 90; c2++) {
          const suffix = `${num}${String.fromCharCode(c1)}${String.fromCharCode(c2)}`;
          const pc = `${prefix} ${suffix}`;

          // Batch validate using postcodes.io (up to 100 at a time)
          // We'll collect and batch below
          postcodes[pc] = null; // placeholder
        }
      }
    }
  }

  // Batch validate postcodes (postcodes.io accepts 100 per request)
  const allCodes = Object.keys(postcodes);
  const validPostcodes = {};
  const batchSize = 100;

  for (let i = 0; i < allCodes.length; i += batchSize) {
    const batch = allCodes.slice(i, i + batchSize);
    const progress = Math.round((i / allCodes.length) * 100);
    process.stdout.write(`\rValidating postcodes... ${progress}% (${Object.keys(validPostcodes).length} valid)`);

    try {
      const result = await new Promise((resolve, reject) => {
        const postData = JSON.stringify({ postcodes: batch });
        const options = {
          hostname: "api.postcodes.io",
          path: "/postcodes",
          method: "POST",
          headers: {
            "Content-Type": "application/json",
            "Content-Length": Buffer.byteLength(postData),
          },
        };

        const req = https.request(options, (res) => {
          let data = "";
          res.on("data", (chunk) => (data += chunk));
          res.on("end", () => {
            try {
              resolve(JSON.parse(data));
            } catch (e) {
              reject(e);
            }
          });
        });
        req.on("error", reject);
        req.write(postData);
        req.end();
      });

      if (result.result) {
        result.result.forEach((r) => {
          if (r.result) {
            validPostcodes[r.query] = {
              lat: r.result.latitude,
              lng: r.result.longitude,
            };
          }
        });
      }
    } catch (err) {
      console.error(`\nBatch error at ${i}: ${err.message}`);
      await sleep(2000);
      i -= batchSize; // retry
    }

    // Small delay to be nice to postcodes.io
    await sleep(200);
  }

  console.log(`\nFound ${Object.keys(validPostcodes).length} valid ZE postcodes`);

  fs.mkdirSync(CACHE_DIR, { recursive: true });
  fs.writeFileSync(POSTCODES_FILE, JSON.stringify(validPostcodes, null, 2));
  return validPostcodes;
}

// --- ScotLIS Fetching ---

function parseConsideration(consideration) {
  if (!consideration) return { price: null, note: "No data" };

  // Extract numeric price
  const priceMatch = consideration.match(/£([\d,]+)/);
  if (priceMatch) {
    return { price: parseInt(priceMatch[1].replace(/,/g, ""), 10), note: null };
  }

  // Non-numeric considerations
  return { price: null, note: consideration };
}

async function fetchPostcode(postcode) {
  const encoded = encodeURIComponent(postcode);
  const url = `https://scotlis.ros.gov.uk/public/bff/land-register/addresses?postcode=${encoded}`;
  const data = await httpsGet(url);

  const results = [];
  const addresses = data?._embedded?.addresses || [];

  for (const addr of addresses) {
    const address = addr.prettyPrint;
    const titles = addr.titles || [];

    for (const title of titles) {
      const { price, note } = parseConsideration(title.consideration);
      results.push({
        address,
        date: title.entryDate || null,
        price,
        note,
        titleNumber: title.titleNumber || null,
      });
    }
  }

  return results;
}

// --- Progress Management ---

function loadProgress() {
  if (fs.existsSync(PROGRESS_FILE)) {
    return JSON.parse(fs.readFileSync(PROGRESS_FILE, "utf8"));
  }
  return { completed: {}, properties: [] };
}

function saveProgress(progress) {
  fs.mkdirSync(CACHE_DIR, { recursive: true });
  fs.writeFileSync(PROGRESS_FILE, JSON.stringify(progress));
}

// --- Export ---

function exportToDataJs(postcodes) {
  const progress = loadProgress();
  const properties = progress.properties;

  console.log(`\nExporting ${properties.length} sales entries...`);

  // Attach lat/lng from postcode
  const entries = properties.map((p) => {
    // Extract postcode from address (last part after final comma)
    const parts = p.address.split(",").map((s) => s.trim());
    const postcode = parts[parts.length - 1];
    const coords = postcodes[postcode] || { lat: null, lng: null };

    const entry = {
      lat: coords.lat,
      lng: coords.lng,
      price: p.price,
      address: p.address,
      date: p.date,
    };

    if (p.note) entry.note = p.note;

    return entry;
  });

  // Filter out entries with no coordinates
  const valid = entries.filter((e) => e.lat !== null && e.lng !== null);
  const noCoords = entries.length - valid.length;
  if (noCoords > 0) {
    console.log(`  Skipped ${noCoords} entries with no coordinates`);
  }

  // Sort by address then date
  valid.sort((a, b) => {
    const addrCmp = a.address.localeCompare(b.address);
    if (addrCmp !== 0) return addrCmp;
    return (a.date || "").localeCompare(b.date || "");
  });

  // Build data.js
  const lines = valid.map((p) => {
    let s = `  { lat: ${p.lat}, lng: ${p.lng}, price: ${p.price}, address: "${p.address}", date: "${p.date}"`;
    if (p.note) s += `, note: "${p.note.replace(/"/g, '\\"')}"`;
    s += " }";
    return s;
  });

  const output = "const properties = [\n" + lines.join(",\n") + "\n];\n";
  fs.writeFileSync(OUTPUT_FILE, output);

  // Stats
  const withPrice = valid.filter((e) => e.price !== null).length;
  const noPrice = valid.filter((e) => e.price === null).length;
  const uniqueAddresses = new Set(valid.map((e) => e.address)).size;

  console.log(`\nExported to data.js:`);
  console.log(`  ${valid.length} total sale entries`);
  console.log(`  ${uniqueAddresses} unique addresses`);
  console.log(`  ${withPrice} with price, ${noPrice} without price`);

  // Show note breakdown
  const notes = {};
  valid.filter((e) => e.note).forEach((e) => {
    notes[e.note] = (notes[e.note] || 0) + 1;
  });
  if (Object.keys(notes).length > 0) {
    console.log(`\n  Non-price considerations:`);
    Object.entries(notes)
      .sort((a, b) => b[1] - a[1])
      .forEach(([note, count]) => {
        console.log(`    "${note}": ${count}`);
      });
  }
}

// --- Main ---

async function main() {
  const args = process.argv.slice(2);

  const postcodes = await fetchAllPostcodes();
  const postcodeList = Object.keys(postcodes);

  if (args.includes("--stats")) {
    const progress = loadProgress();
    const done = Object.keys(progress.completed).length;
    console.log(`Progress: ${done}/${postcodeList.length} postcodes fetched`);
    console.log(`Properties found: ${progress.properties.length}`);
    const remaining = postcodeList.length - done;
    const avgDelay = 20.5; // midpoint of 18-23
    const etaMinutes = Math.round((remaining * avgDelay) / 60);
    console.log(`Estimated time remaining: ${etaMinutes} minutes (${(etaMinutes / 60).toFixed(1)} hours)`);
    return;
  }

  if (args.includes("--export")) {
    exportToDataJs(postcodes);
    return;
  }

  // Fetch mode
  const progress = loadProgress();
  const remaining = postcodeList.filter((pc) => !progress.completed[pc]);

  console.log(`\n${postcodeList.length} total postcodes, ${remaining.length} remaining`);
  console.log(`${progress.properties.length} properties found so far`);
  console.log(`Rate limit: 18-23s between requests\n`);

  if (remaining.length === 0) {
    console.log("All postcodes already fetched! Use --export to generate data.js");
    return;
  }

  const startTime = Date.now();
  let fetchedThisSession = 0;

  for (const postcode of remaining) {
    const idx = postcodeList.indexOf(postcode) + 1;
    const elapsed = (Date.now() - startTime) / 1000;
    const rate = fetchedThisSession > 0 ? elapsed / fetchedThisSession : 20.5;
    const eta = Math.round(((remaining.length - fetchedThisSession) * rate) / 60);

    process.stdout.write(
      `\r[${idx}/${postcodeList.length}] ${postcode} — ${progress.properties.length} properties — ETA: ${eta}m    `
    );

    try {
      const results = await fetchPostcode(postcode);
      progress.properties.push(...results);
      progress.completed[postcode] = { count: results.length, time: new Date().toISOString() };
      fetchedThisSession++;

      // Save after each postcode
      saveProgress(progress);
    } catch (err) {
      console.error(`\nError fetching ${postcode}: ${err.message}`);
      // Save progress and continue
      saveProgress(progress);

      // Wait longer on errors (possible rate limit)
      await sleep(30000);
      continue;
    }

    // Random delay
    const delay = randomDelay();
    await sleep(delay);
  }

  console.log(`\n\nDone! Fetched ${fetchedThisSession} postcodes this session.`);
  console.log(`Total properties: ${progress.properties.length}`);
  console.log(`\nRun with --export to generate data.js`);
}

main().catch((err) => {
  console.error("Fatal error:", err);
  process.exit(1);
});

File: fetch_data.py
#!/usr/bin/env python3
"""
Fetch Shetland property price data from ScotLIS and geocode via postcodes.io.
Outputs data.js for the map application.

Resumes from cached postcode list and already-fetched results if available.
"""

import json
import os
import random
import re
import time
import sys
import requests

SCOTLIS_DELAY_MIN = 18  # Random delay between 18-33s to mimic human browsing
SCOTLIS_DELAY_MAX = 33
POSTCODES_IO_DELAY = 0.15
CACHE_DIR = "cache"
POSTCODES_CACHE = os.path.join(CACHE_DIR, "postcodes.json")
PROGRESS_CACHE = os.path.join(CACHE_DIR, "progress.json")

SESSION = requests.Session()
SESSION.headers.update({
    "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36",
    "Accept": "application/json, text/plain, */*",
    "Accept-Language": "en-GB,en;q=0.9",
    "Referer": "https://scotlis.ros.gov.uk/",
})


def ensure_cache_dir():
    os.makedirs(CACHE_DIR, exist_ok=True)


def get_all_ze_postcodes():
    """Enumerate all valid ZE postcodes via postcodes.io bulk validation."""
    if os.path.exists(POSTCODES_CACHE):
        with open(POSTCODES_CACHE) as f:
            cached = json.load(f)
        print(f"  Loaded {len(cached)} postcodes from cache")
        return cached

    letters = "ABCDEFGHJKLMNPQRSTUVWXYZ"
    valid = {}

    for outcode in ["ZE1", "ZE2", "ZE3"]:
        for sector in range(10):
            candidates = [f"{outcode} {sector}{l1}{l2}" for l1 in letters for l2 in letters]

            for i in range(0, len(candidates), 100):
                batch = candidates[i : i + 100]
                try:
                    resp = SESSION.post(
                        "https://api.postcodes.io/postcodes",
                        json={"postcodes": batch},
                        timeout=15,
                    )
                    data = resp.json()
                    for item in data.get("result", []):
                        if item.get("result"):
                            r = item["result"]
                            valid[r["postcode"]] = {
                                "lat": r["latitude"],
                                "lng": r["longitude"],
                            }
                except Exception as e:
                    print(f"  Error validating batch: {e}", file=sys.stderr)
                time.sleep(POSTCODES_IO_DELAY)

            print(f"  {outcode} {sector}xx: {len(valid)} valid postcodes so far")

    # Cache for reuse
    ensure_cache_dir()
    with open(POSTCODES_CACHE, "w") as f:
        json.dump(valid, f)
    print(f"  Cached postcodes to {POSTCODES_CACHE}")

    return valid


def fetch_scotlis_properties(postcode, max_retries=3):
    """Fetch properties for a postcode from ScotLIS with retry logic."""
    url = "https://scotlis.ros.gov.uk/public/bff/land-register/addresses"

    for attempt in range(max_retries):
        try:
            resp = SESSION.get(url, params={"postcode": postcode.lower()}, timeout=15)

            if resp.status_code == 202 or not resp.text.strip():
                # Rate limited - back off aggressively
                wait = 60 * (attempt + 1)  # 1min, 2min, 3min
                print(f"  Rate limited on {postcode}, waiting {wait}s...", file=sys.stderr)
                time.sleep(wait)
                continue

            if resp.status_code != 200:
                print(f"  HTTP {resp.status_code} for {postcode}", file=sys.stderr)
                return []

            data = resp.json()
        except Exception as e:
            print(f"  Error fetching {postcode} (attempt {attempt+1}): {e}", file=sys.stderr)
            time.sleep(3)
            continue

        if "_embedded" not in data:
            return []

        properties = []
        for addr in data["_embedded"].get("addresses", []):
            address = addr.get("prettyPrint", "")
            for title in addr.get("titles", []):
                consideration = title.get("consideration", "")
                entry_date = title.get("entryDate", "")

                price_match = re.search(r"£([\d,]+)", consideration)
                if not price_match:
                    continue

                price = int(price_match.group(1).replace(",", ""))
                if price < 5000:
                    continue

                properties.append({
                    "address": address,
                    "price": price,
                    "entryDate": entry_date,
                    "postcode": postcode,
                })

        return properties

    print(f"  Failed after {max_retries} retries for {postcode}", file=sys.stderr)
    return []


def load_progress():
    """Load previously fetched results so we can resume."""
    if os.path.exists(PROGRESS_CACHE):
        with open(PROGRESS_CACHE) as f:
            return json.load(f)
    return {"fetched_postcodes": [], "properties": []}


def save_progress(progress):
    """Save progress to disk."""
    ensure_cache_dir()
    with open(PROGRESS_CACHE, "w") as f:
        json.dump(progress, f)


def main():
    print("=== Shetland Property Price Data Fetcher ===\n")
    ensure_cache_dir()

    # Step 1: Get all valid ZE postcodes with coordinates
    print("Step 1: Finding all valid Shetland postcodes...")
    valid_postcodes = get_all_ze_postcodes()
    print(f"\n  Found {len(valid_postcodes)} valid postcodes\n")

    # Step 2: Fetch property data from ScotLIS (with resume support)
    print(f"Step 2: Fetching property data from ScotLIS...")
    progress = load_progress()
    already_fetched = set(progress["fetched_postcodes"])
    all_properties = progress["properties"]

    postcodes_list = sorted(valid_postcodes.keys())
    remaining = [pc for pc in postcodes_list if pc not in already_fetched]

    if already_fetched:
        print(f"  Resuming: {len(already_fetched)} already done, {len(remaining)} remaining")
        print(f"  {len(all_properties)} properties found so far")

    with_data = sum(1 for pc in already_fetched if any(p["postcode"] == pc for p in all_properties))

    for i, pc in enumerate(remaining):
        props = fetch_scotlis_properties(pc)
        if props:
            with_data += 1
            coords = valid_postcodes[pc]
            for p in props:
                p["lat"] = coords["lat"]
                p["lng"] = coords["lng"]
            all_properties.extend(props)

        progress["fetched_postcodes"].append(pc)
        progress["properties"] = all_properties

        total_done = len(already_fetched) + i + 1
        if (i + 1) % 50 == 0 or i == len(remaining) - 1:
            save_progress(progress)
            print(
                f"  Processed {total_done}/{len(postcodes_list)} postcodes | "
                f"{with_data} with data | {len(all_properties)} properties with prices"
            )
        time.sleep(random.uniform(SCOTLIS_DELAY_MIN, SCOTLIS_DELAY_MAX))

    print(f"\n  Total properties with prices: {len(all_properties)}")

    # Step 3: Deduplicate by address (keep most recent entry)
    print("\nStep 3: Deduplicating...")
    by_address = {}
    for p in all_properties:
        key = p["address"]
        if key not in by_address or p["entryDate"] > by_address[key]["entryDate"]:
            by_address[key] = p

    unique = sorted(by_address.values(), key=lambda p: p["address"])
    print(f"  {len(unique)} unique properties after dedup")

    # Step 4: Write data.js
    print("\nStep 4: Writing data.js...")
    lines = ["const properties = ["]
    for p in unique:
        addr = p["address"].replace("\\", "\\\\").replace('"', '\\"')
        lines.append(
            f'  {{ lat: {p["lat"]}, lng: {p["lng"]}, price: {p["price"]}, '
            f'address: "{addr}", date: "{p["entryDate"]}" }},'
        )
    lines.append("];")

    with open("data.js", "w") as f:
        f.write("\n".join(lines) + "\n")
    print(f"  Wrote {len(unique)} properties to data.js")

    # Also save raw JSON
    with open("data.json", "w") as f:
        json.dump(unique, f, indent=2)
    print(f"  Wrote raw data to data.json")

    print("\nDone!")


if __name__ == "__main__":
    main()
